README.md
# NAME

JSON::XS::ByteString - A more predictable and convenient XS implementation for JSON

# SYNOPSIS

    use JSON::XS::ByteString qw(encode_json encode_json_unblessed decode_json decode_json_safe);

    $json_string = encode_json($perl_data);
    $json_string_pretty = encode_json_pretty($perl_data);
    $perl_data = decode_json($json_string);
    $perl_data = decode_json($json_string, 1); # die if $json_string is not valid JSON string

    $json_string = encode_json_unblessed($perl_data);
    $json_string_pretty = encode_json_unblessed_pretty($perl_data);
        # the same behavior as encode_json
        #  but encode blessed references as reference strings,
        #  like 'Object=HASH(0xffffffff)'

# DESCRIPTION

This module is a XS implementation for JSON. It provide a more predictable behavior than [JSON::XS](https://metacpan.org/pod/JSON%3A%3AXS) by always producing strings in JSON for normal scalars.
And you can force it to produce numbers in JSON by putting references to numbers.

All the string data are treated as UTF-8 octets and just copy them in and out directly, except `"`, `\` and characters that `ord($char) < 32`

`decode_json` will return an undef without exceptions with invalid json string.

# DESIGN CONSIDERATION

## I didn't transfer the numeric value from `json_decode` back to string values

Because in the pure Perl world, there's insignificant difference between numeric or string.
So I think we don't need to do it since the result will be used in Perl.

## I didn't transfer the numeric value from `json_decode` back to reference values

Let `json_decode` preserve the identical structure as it received.

# FUNCTIONS

## $json\_string = encode\_json($perl\_data) / encode\_json\_pretty($perl\_data)

Get a JSON string from a perl data structure. Treat blessed objects as normal references.

## $json\_string = encode\_json\_unblessed($perl\_data) / encode\_json\_unblessed\_pretty($perl\_data)

Get a JSON string from a perl data structure. Treat blessed objects as strings (such as `'Object=HASH(0xffffffff)'`)

## $perl\_data = decode\_json($json\_string, $warn2die=0)

Get the perl data structure back from a JSON string.

If the given string is not a valid JSON string, it will return an `undef`.
If the `$warn2die` is false or not specified, this function will not die but warns an offset where it encountered the unrecognized character.
If the `$warn2die` is true, this function will die with the error message which is identical to the warning one.

## $perl\_data = decode\_json\_safe($json\_string)

The same as `decode_json` except that `decode_json_safe` will not warn at all.

# SEE ALSO

[JSON::XS](https://metacpan.org/pod/JSON%3A%3AXS)

This mod's github repository [https://github.com/CindyLinz/Perl-JSON-XS-ByteString](https://github.com/CindyLinz/Perl-JSON-XS-ByteString)

# AUTHOR

Cindy Wang (CindyLinz)

# COPYRIGHT AND LICENSE

Copyright (C) 2014-2021 by Cindy Wang (CindyLinz)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension JSON::XS::ByteString.

1.7.0 2022.11.5
    Bug fix for encode_json_pretty and encode_json_unblessed_pretty

1.6.0 2022.11.3
    Add encode_json_pretty and encode_json_unblessed_pretty

1.5.0 2021.6.19
    Remove a debug message in `decode_json`

1.4.0 2020.4.18
    decode_json with warn2die option

1.3.1 2017.7.25
    Compatible with perl 5.16 or below

1.3.0 2017.7.25
    Buf fix for decoding identity (true, false, null).
    Skip UTF-8 BOM automatically

1.2.0 2017.7.21
    Bug fix for unsigned STRLEN

1.1.0 2017.7.21
    Parse hash keys and value strings with different parser.
    Return undef when encountering unrecgonized character series.

1.0.6 2017.5.12
    Degenerate looped data point to NULL when generating JSON strings.

1.0.5 2017.4.26
    Avoid "</" by "<\/" in the generated JSON strings.

1.0.4 2017.4.26
    Deal with escaped slash when decoding.

1.0.3 2017.4.25
    A missing file.

1.0.2 2017.4.24
    Avoid warnings on clang

1.0.1 2017.4.24
    Edit documentation

1.0.0 2017.4.24
    Reimplement. Remove the dependency to JSON::XS.

0.15.3 2016.4.14
    Fix floating-point values bug on some perl version.

0.15.2 2016.3.25
    true -> TURE
    false -> FALSE

0.15.1 2016.3.24
    Restore origin bug fix

0.15.0 2016.3.24
    Deal with blessed references

0.14.0 2014.9.12
    Add numeric semantics [EXPERIMENTAL]

0.13.5 2014.8.27
    To fit JSON::XS' stricter acceptable codepoint

0.13.4 2014.8.27
    Fix malform UTF8 octet shadow buffer's wrong end position.

0.13.3 2014.8.13
    To compatible with Perl 5.8.5

0.13.2 2014.8.12
    Use older form of version string, to compatible with Perl 5.8

0.13.1 2014.8.1
    Add some syntax additions that makes clang happier...

0.13.0 2014.8.1
    Deal with malform UTF8 octet (except in hash key)

0.12.0 2014.7.9
    Deal with UTF8/wide char hash key

0.11.0 2014.7.9
    Detect Perl undef more correctly

0.10.0 2014.7.5
    Preserve Perl undef <-> JSON null

0.9.0  2014.5.23
    The first release ^_^


Makefile.PL
use 5.008;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'JSON::XS::ByteString',
    VERSION_FROM      => 'lib/JSON/XS/ByteString.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/JSON/XS/ByteString.pm', # retrieve abstract from module
       AUTHOR         => 'cindy <cindy@cpan.org>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => ' -g -O0 -std=gnu99', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


encode_gen.h
// vim: filetype=xs

STRLEN CONCAT(estimate_, NAME)(SV * sv, int depth){
    if( sv!=NULL ){
        if( SvROK(sv) && (!UNBLESSED || !sv_isobject(sv)) ){
            SV * rvs = SvRV(sv);
            for(int i=visited_p-1; i>=0; --i)
                if( visited[i] == rvs )
                    return 4;
            reserve_visited_capacity();
            visited[visited_p++] = rvs;
            switch( SvTYPE(rvs) ){
                case SVt_PVAV: {
                    AV * av = (AV*) rvs;
                    SSize_t n = av_len(av);
                    if( n<0 ){
                        --visited_p;
                        return 2;
                    }

                    STRLEN len = 2 + n;
                    SV ** elems = AvARRAY(av);
                    for(int i=0; i<=n; ++i)
                        len += CONCAT(estimate_, NAME)(elems[i], depth+1);
                    if( PRETTY && n>0 )
                        len += (n+1) * (depth*2+2);
                    --visited_p;
                    return len;
                }
                case SVt_PVHV: {
                    HV * hv = (HV*) rvs;

                    STRLEN len = 1;
                    hv_iterinit(hv);
                    for(HE * entry = hv_iternext(hv); entry; entry = hv_iternext(hv)){
                        len += 2; // : ,

                        I32 keylen;
                        char * key = hv_iterkey(entry, &keylen);
                        len += estimate_str((unsigned char*)key, (STRLEN) keylen);

                        SV * val = hv_iterval(hv, entry);
                        len += CONCAT(estimate_, NAME)(val, depth+1);
                        if( PRETTY )
                            len += depth*2 + 3;
                    }
                    if( len==1 )
                        ++len;

                    --visited_p;
                    return len;
                }
                default:
                    --visited_p;
                    break;
            }
            if( SvTYPE(rvs) < SVt_PVAV ){
                NV nv = SvNV(rvs);
                IV iv;
                STRLEN len=0;
                if( !Perl_isnan(nv) && nv == (NV)(iv = (IV) nv) ){
                    if( iv == 0 )
                        return 1;
                    if( iv < 0 )
                        ++len;
                    while( iv ){
                        ++len;
                        iv /= 10;
                    }
                }
                else{
                    char buffer[100];
                    snprintf(buffer, 100, "%g", (double) nv);
                    while( buffer[len] )
                        ++len;
                }
                return len;
            }
        }
        if( SvOK(sv) ){
            STRLEN len;
            char * str = SvPV(sv, len);
            return estimate_str((unsigned char*)str, len);
        }
    }
    return 4;
}

unsigned char * CONCAT(encode_, NAME)(unsigned char * buffer, SV * sv, int depth){
    if( sv!=NULL ){
        if( SvROK(sv) && (!UNBLESSED || !sv_isobject(sv)) ){
            SV * rvs = SvRV(sv);
            for(int i=visited_p-1; i>=0; --i)
                if( visited[i] == rvs )
                    goto DEGENERATE;
            reserve_visited_capacity();
            visited[visited_p++] = rvs;
            switch( SvTYPE(rvs) ){
                case SVt_PVAV: {
                    *buffer++ = '[';

                    AV * av = (AV*) rvs;
                    SSize_t n = av_len(av);
                    if( n>=0 ){
                        SV ** elems = AvARRAY(av);
                        for(int i=0; i<n; ++i){
                            if( PRETTY )
                                *buffer++ = ' ';
                            buffer = CONCAT(encode_, NAME)(buffer, elems[i], depth+1);
                            if( PRETTY ){
                                *buffer++ = '\n';
                                for(int i=0; i<depth; ++i){
                                    *buffer++ = ' ';
                                    *buffer++ = ' ';
                                }
                            }
                            *buffer++ = ',';
                        }
                        if( PRETTY && n>0 )
                            *buffer++ = ' ';
                        buffer = CONCAT(encode_, NAME)(buffer, elems[n], depth+1);
                        if( PRETTY && n>0 ){
                            *buffer++ = '\n';
                            for(int i=0; i<depth; ++i){
                                *buffer++ = ' ';
                                *buffer++ = ' ';
                            }
                        }
                    }

                    *buffer++ = ']';
                    --visited_p;
                    return buffer;
                }
                case SVt_PVHV: {
                    *buffer++ = '{';
                    HV * hv = (HV*) rvs;

                    hv_iterinit(hv);
                    for(HE * entry = hv_iternext(hv); entry; entry = hv_iternext(hv)){
                        if( PRETTY )
                            *buffer++ = ' ';

                        I32 keylen;
                        char * key = hv_iterkey(entry, &keylen);
                        buffer = encode_str(buffer, (unsigned char*)key, (STRLEN) keylen);

                        *buffer++ = ':';
                        if( PRETTY )
                            *buffer++ = ' ';

                        SV * val = hv_iterval(hv, entry);
                        buffer = CONCAT(encode_, NAME)(buffer, val, depth+1);

                        if( PRETTY ){
                            *buffer++ = '\n';
                            for(int i=0; i<depth; ++i){
                                *buffer++ = ' ';
                                *buffer++ = ' ';
                            }
                        }
                        *buffer++ = ',';
                    }

                    if( *(buffer-1) == '{' )
                        *buffer++ = '}';
                    else
                        *(buffer-1) = '}';
                    --visited_p;
                    return buffer;
                }
                default:
                    --visited_p;
                    break;
            }
            if( SvTYPE(rvs) < SVt_PVAV ){
                NV nv = SvNV(rvs);
                IV iv;
                if( !Perl_isnan(nv) && nv == (NV)(iv = (IV) nv) ){
                    if( -59074 <= iv && iv <= 59074 ){
                        // (stolen from JSON::XS)
                        // optimise the "small number case"
                        // code will likely be branchless and use only a single multiplication
                        // works for numbers up to 59074
                        U32 u;
                        char digit, nz = 0;
                        if( iv < 0 ){
                            *buffer++ = '-';
                            u = -iv;
                        }
                        else
                            u = iv;
                        // convert to 4.28 fixed-point representation
                        u = u * ((0xfffffff + 10000) / 10000); // 10**5, 5 fractional digits

                        // now output digit by digit, each time masking out the integer part
                        // and multiplying by 5 while moving the decimal point one to the right,
                        // resulting in a net multiplication by 10.
                        // we always write the digit to memory but conditionally increment
                        // the pointer, to enable the use of conditional move instructions.
                        digit = u >> 28; *buffer = digit + '0'; buffer += (nz = nz || digit); u = (u & 0xfffffffUL) * 5;
                        digit = u >> 27; *buffer = digit + '0'; buffer += (nz = nz || digit); u = (u & 0x7ffffffUL) * 5;
                        digit = u >> 26; *buffer = digit + '0'; buffer += (nz = nz || digit); u = (u & 0x3ffffffUL) * 5;
                        digit = u >> 25; *buffer = digit + '0'; buffer += (nz = nz || digit); u = (u & 0x1ffffffUL) * 5;
                        digit = u >> 24; *buffer = digit + '0'; buffer += 1; // correctly generate '0'
                    }
                    else{
                        snprintf((char*)buffer, 100, "%lld", (long long)iv);
                        while( *buffer )
                            ++buffer;
                    }
                }
                else{
                    snprintf((char*)buffer, 100, "%g", (double) nv);
                    while( *buffer )
                        ++buffer;
                }
                return buffer;
            }
        }
        if( SvOK(sv) ){
            STRLEN len;
            char * str = SvPV(sv, len);
            return encode_str(buffer, (unsigned char*)str, len);
        }
    }
    DEGENERATE:
    *buffer++ = 'n';
    *buffer++ = 'u';
    *buffer++ = 'l';
    *buffer++ = 'l';
    return buffer;
}



lib/JSON/XS/ByteString.pm
package JSON::XS::ByteString;

use 5.008;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);
our @EXPORT_OK = qw(encode_json encode_json_pretty encode_json_unblessed encode_json_unblessed_pretty decode_json decode_json_safe);
our $VERSION = 1.007000;

require XSLoader;
XSLoader::load('JSON::XS::ByteString', $VERSION);

=head1 NAME

JSON::XS::ByteString - A more predictable and convenient XS implementation for JSON

=head1 SYNOPSIS

    use JSON::XS::ByteString qw(encode_json encode_json_unblessed decode_json decode_json_safe);

    $json_string = encode_json($perl_data);
    $json_string_pretty = encode_json_pretty($perl_data);
    $perl_data = decode_json($json_string);
    $perl_data = decode_json($json_string, 1); # die if $json_string is not valid JSON string

    $json_string = encode_json_unblessed($perl_data);
    $json_string_pretty = encode_json_unblessed_pretty($perl_data);
        # the same behavior as encode_json
        #  but encode blessed references as reference strings,
        #  like 'Object=HASH(0xffffffff)'

=head1 DESCRIPTION

This module is a XS implementation for JSON. It provide a more predictable behavior than L<JSON::XS> by always producing strings in JSON for normal scalars.
And you can force it to produce numbers in JSON by putting references to numbers.

All the string data are treated as UTF-8 octets and just copy them in and out directly, except C<">, C<\> and characters that C<< ord($char) < 32 >>

C<decode_json> will return an undef without exceptions with invalid json string.

=head1 DESIGN CONSIDERATION

=head2 I didn't transfer the numeric value from C<json_decode> back to string values

Because in the pure Perl world, there's insignificant difference between numeric or string.
So I think we don't need to do it since the result will be used in Perl.

=head2 I didn't transfer the numeric value from C<json_decode> back to reference values

Let C<json_decode> preserve the identical structure as it received.

=head1 FUNCTIONS

=head2 $json_string = encode_json($perl_data) / encode_json_pretty($perl_data)

Get a JSON string from a perl data structure. Treat blessed objects as normal references.

=head2 $json_string = encode_json_unblessed($perl_data) / encode_json_unblessed_pretty($perl_data)

Get a JSON string from a perl data structure. Treat blessed objects as strings (such as C<'Object=HASH(0xffffffff)'>)

=head2 $perl_data = decode_json($json_string, $warn2die=0)

Get the perl data structure back from a JSON string.

If the given string is not a valid JSON string, it will return an C<undef>.
If the C<$warn2die> is false or not specified, this function will not die but warns an offset where it encountered the unrecognized character.
If the C<$warn2die> is true, this function will die with the error message which is identical to the warning one.

=head2 $perl_data = decode_json_safe($json_string)

The same as C<decode_json> except that C<decode_json_safe> will not warn at all.

=head1 SEE ALSO

L<JSON::XS>

This mod's github repository L<https://github.com/CindyLinz/Perl-JSON-XS-ByteString>

=head1 AUTHOR

Cindy Wang (CindyLinz)

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2021 by Cindy Wang (CindyLinz)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
